### PR TITLE
Relational: Translate SelectMany

### LINKED ISSUE #12567: Query: NRE for query with top level DefaultIfEmpty and projecting a constant
query using Northwind model:

```cs
from e in ctx.Employees.Where(c => c.EmployeeID == NonExistentID).DefaultIfEmpty()
select "Foo";
```

query plan:

```
(QueryContext queryContext) => IEnumerable<string> _InterceptExceptions(
|__ source: IEnumerable<string> DefaultIfEmpty(IEnumerable<string> _DefaultIfEmptyShapedQuery(
|   |__ queryContext: queryContext, 
|   |__ shaperCommandContext: SelectExpression: 
|   |       SELECT 1
|   |       FROM (
|   |           SELECT NULL AS [empty]
|   |       ) AS [empty]
|   |       LEFT JOIN (
|   |           SELECT [c].*
|   |           FROM [Employees] AS [c]
|   |           WHERE [c].[EmployeeID] = -1
|   |       ) AS [t] ON 1 = 1, 
|   |__ shaper: (QueryContext queryContext | ValueBuffer c) => "Foo")), 
|__ contextType: TestModels.Northwind.NorthwindRelationalContext, 
|__ logger: DiagnosticsLogger<Query>, 
|__ queryContext: Unhandled parameter: queryContext)
```

Exception:

```
System.NullReferenceException : Object reference not set to an instance of an object.
	at Microsoft.EntityFrameworkCore.Storage.ValueBuffer.get_Count()
	at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_DefaultIfEmptyShapedQuery>d__11`1.MoveNext()
	at System.Linq.Enumerable.<DefaultIfEmptyIterator>d__93`1.MoveNext()
	at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext()
	at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
	at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
	at Microsoft.EntityFrameworkCore.Query.SimpleQueryTestBase`1.Default_if_empty_top_level(Boolean isAsync)
	at Microsoft.EntityFrameworkCore.Query.SimpleQuerySqlServerTest.<>n__7(Boolean isAsync)
	at Microsoft.EntityFrameworkCore.Query.SimpleQuerySqlServerTest.<Default_if_empty_top_level>d__14.MoveNext()
	--- End of stack trace from previous location where exception was thrown ---
	at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	--- End of stack trace from previous location where exception was thrown ---
	at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	--- End of stack trace from previous location where exception was thrown ---
	at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

### LINKED ISSUE #12572: Query: compilation error for query with top level DefaultIfEmpty that takes an argument, and projecting constant on top
query:

```
from e in es.Where(c => c.EmployeeID == NonExistentID).DefaultIfEmpty(new Employee())
select 42
```

Exception:

```
Expression of type 'Microsoft.EntityFrameworkCore.TestModels.Northwind.Employee' cannot be used for parameter of type 'System.Int32' of method 'System.Collections.Generic.IEnumerable`1[System.Int32] DefaultIfEmpty[Int32](System.Collections.Generic.IEnumerable`1[System.Int32], Int32)'
	at System.Linq.Expressions.Expression.ValidateOneArgument(MethodBase method, ExpressionType nodeKind, Expression arg, ParameterInfo pi)
	at System.Linq.Expressions.Expression.Call(MethodInfo method, Expression arg0, Expression arg1)
	at Microsoft.EntityFrameworkCore.Query.ResultOperatorHandler.HandleDefaultIfEmpty(EntityQueryModelVisitor entityQueryModelVisitor, DefaultIfEmptyResultOperator defaultIfEmptyResultOperator, QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.ResultOperatorHandler.<>c.<.cctor>b__33_7(EntityQueryModelVisitor v, ResultOperatorBase r, QueryModel q)
	at Microsoft.EntityFrameworkCore.Query.ResultOperatorHandler.HandleResultOperator(EntityQueryModelVisitor entityQueryModelVisitor, ResultOperatorBase resultOperator, QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.Internal.RelationalResultOperatorHandler.HandlerContext.EvalOnClient(Boolean requiresClientResultOperator)
	at Microsoft.EntityFrameworkCore.Query.Internal.RelationalResultOperatorHandler.HandleDefaultIfEmpty(HandlerContext handlerContext)
	at Microsoft.EntityFrameworkCore.Query.Internal.RelationalResultOperatorHandler.HandleResultOperator(EntityQueryModelVisitor entityQueryModelVisitor, ResultOperatorBase resultOperator, QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitResultOperator(ResultOperatorBase resultOperator, QueryModel queryModel, Int32 index)
	at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitResultOperator(ResultOperatorBase resultOperator, QueryModel queryModel, Int32 index)
	at Remotion.Linq.Clauses.ResultOperatorBase.Accept(IQueryModelVisitor visitor, QueryModel queryModel, Int32 index)
	at Remotion.Linq.QueryModelVisitorBase.VisitResultOperators(ObservableCollection`1 resultOperators, QueryModel queryModel)
	at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel)
	--- End of stack trace from previous location where exception was thrown ---
	at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass13_0`1.<Execute>b__0()
	at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
	at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
	at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
	at Remotion.Linq.QueryableBase`1.GetEnumerator()
	at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
	at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
	at Microsoft.EntityFrameworkCore.Query.SimpleQueryTestBase`1.Default_if_empty_top_level_arg(Boolean isAsync)
	at Microsoft.EntityFrameworkCore.Query.SimpleQuerySqlServerTest.<>n__9(Boolean isAsync)
	at Microsoft.EntityFrameworkCore.Query.SimpleQuerySqlServerTest.<Default_if_empty_top_level_arg>d__16.MoveNext()
	--- End of stack trace from previous location where exception was thrown ---
	at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	--- End of stack trace from previous location where exception was thrown ---
	at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
	--- End of stack trace from previous location where exception was thrown ---
	at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
	at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

### LINKED ISSUE #12872: Query: compilation error (Operation is not valid due to the current state of the object) for query with DefaultIfEmpty joined with itself
query:

```
                var query =
                    (from e in ctx.Employees.Where(c => c.EmployeeID > 0).DefaultIfEmpty()
                     select e).Join(from e in ctx.Employees.Where(c => c.EmployeeID > 0).DefaultIfEmpty()
                                    select e, o => o, i => i, (o, i) => o);

                var result = query.ToList();

```

exception:

```
Operation is not valid due to the current state of the object.
	at System.Linq.Expressions.MethodCallExpression1.System.Linq.Expressions.IArgumentProvider.GetArgument(Int32 index)
	at System.Linq.Expressions.ListArgumentProvider.get_Item(Int32 index)
	at System.Collections.ObjectModel.ReadOnlyCollection`1.get_Item(Int32 index)
	at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.TryFlattenJoin(JoinClause joinClause, QueryModel queryModel, Int32 index, Int32 previousProjectionCount, ParameterExpression previousParameter, Dictionary`2 previousMapping)
	at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitJoinClause(JoinClause joinClause, QueryModel queryModel, Int32 index)
	at Remotion.Linq.Clauses.JoinClause.Accept(IQueryModelVisitor visitor, QueryModel queryModel, Int32 index)
	at Remotion.Linq.QueryModelVisitorBase.VisitBodyClauses(ObservableCollection`1 bodyClauses, QueryModel queryModel)
	at Remotion.Linq.QueryModelVisitorBase.VisitQueryModel(QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.RelationalQueryModelVisitor.VisitQueryModel(QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Query.EntityQueryModelVisitor.CreateQueryExecutor[TResult](QueryModel queryModel)
	at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](QueryModel queryModel)
	--- End of stack trace from previous location where exception was thrown ---
	at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](Expression query, IQueryModelGenerator queryModelGenerator, IDatabase database, IDiagnosticsLogger`1 logger, Type contextType)
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass13_0`1.<Execute>b__0()
	at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
	at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
	at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
	at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
	at Remotion.Linq.QueryableBase`1.GetEnumerator()
	at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
	at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
```

### LINKED ISSUE #15081: QueryRewrite: convert correlated SelectMany to JOINs
query:

```cs
entities.SelectMany(l1 => l1.OneToMany_Required1)
```

expected:

```sql
SELECT [l1.OneToMany_Required1].[Id], [l1.OneToMany_Required1].[Date], [l1.OneToMany_Required1].[Level1_Optional_Id], [l1.OneToMany_Required1].[Level1_Required_Id], [l1.OneToMany_Required1].[Name], [l1.OneToMany_Required1].[OneToMany_Optional_Inverse2Id], [l1.OneToMany_Required1].[OneToMany_Optional_Self_Inverse2Id], [l1.OneToMany_Required1].[OneToMany_Required_Inverse2Id], [l1.OneToMany_Required1].[OneToMany_Required_Self_Inverse2Id], [l1.OneToMany_Required1].[OneToOne_Optional_PK_Inverse2Id], [l1.OneToMany_Required1].[OneToOne_Optional_Self2Id]
FROM [LevelOne] AS [l1]
INNER JOIN [LevelTwo] AS [l1.OneToMany_Required1] ON [l1].[Id] = [l1.OneToMany_Required1].[OneToMany_Required_Inverse2Id]
```

current:

```sql
SELECT [l].[Id], [l].[Date], [l].[Level1_Optional_Id], [l].[Level1_Required_Id], [l].[Name], [l].[OneToMany_Optional_Inverse2Id], [l].[OneToMany_Optional_Self_Inverse2Id], [l].[OneToMany_Required_Inverse2Id], [l].[OneToMany_Required_Self_Inverse2Id], [l].[OneToOne_Optional_PK_Inverse2Id], [l].[OneToOne_Optional_Self2Id]
FROM [LevelOne] AS [l1]
CROSS JOIN [LevelTwo] AS [l]
WHERE ([l1].[Id] = [l].[OneToMany_Required_Inverse2Id])
```

### LINKED ISSUE #15711: Queryable methods translation progress
Relational:
- [x] All
- [x] Any
- [x] Average
- [x] Cast
- [x] Concat
- [x] Contains
- [x] Count
- [x] DefaultIfEmpty
- [x] Distinct
~- [ ] ElementAt (#17066)~
~- [ ] ElementAtOrDefault (#17066)~
- [x] Except
- [x] First
- [x] FirstOrDefault
- [x] GroupBy
~- [ ] GroupJoin (#17068)~
- [x] Intersect
- [x] Join
- [x] LeftJoin
- [x] Last
- [x] LastOrDefault
- [x] LongCount
- [x] Max
- [x] Min
- [x] OfType
- [x] OrderBy
- [x] OrderByDescending
- [x] Reverse
- [x] Select
- [x] SelectMany
- [x] Single
- [x] SingleOrDefault
- [x] Skip
~- [ ] SkipWhile (#17065)~
- [x] Sum
- [x] Take
~- [ ] TakeWhile (#17065)~
- [x] ThenBy
- [x] ThenByDescending
- [x] Union
- [x] Where

InMemory:
- [x] All
- [x] Any
- [x] Average
~- [ ] Cast (#16963)~
- [x] Concat
- [x] Contains
- [x] Count
~- [ ] DefaultIfEmpty (#16963)~
- [x] Distinct
~- [ ] ElementAt (#17066)~
~- [ ] ElementAtOrDefault (#17066)~
- [x] Except
- [x] First
- [x] FirstOrDefault
~- [ ] GroupBy (#16963)~
~- [ ] GroupJoin (#17068)~
- [x] Intersect
- [x] Join
- [x] LeftJoin
- [x] Last
- [x] LastOrDefault
- [x] LongCount
- [x] Max
- [x] Min
~- [ ] OfType (#16963)~
- [x] OrderBy
- [x] OrderByDescending
~- [ ] Reverse (#16963)~
- [x] Select
~- [ ] SelectMany (#16963)~
- [x] Single
- [x] SingleOrDefault
- [x] Skip
~- [ ] SkipWhile (#17065)~
- [x] Sum
- [x] Take
~- [ ] TakeWhile (#17065)~
- [x] ThenBy
- [x] ThenByDescending
- [x] Union
- [x] Where

Cosmos:

Unknowns:
~- [ ] Aggregate (Not Needed)~
- [x] AsQueryable
~- [ ] SequenceEqual (Not Needed)~
~- [ ] Zip (Not Needed)~
~- [ ] SkipLast (#17065)~
~- [ ] TakeLast (#17065)~

### LINKED ISSUE #16330: Query: FirstOrDefault on collection projection may fail to find correlation
```C#
var actual = ctx.Gears.OrderBy(g => g.Nickname).Select(g => g.Weapons).FirstOrDefault();
                
```
Conditions
- FirstOrDefault on top level which causes a pushdown
- The correlation predicate is not part of identifier for outer
- The correlation predicate is not being projected out from outer.

This causes it to lose reference when pushing down and then fail to identify correlation predicate.

### LINKED ISSUE #16989: SelectMany with collection navigation throws
Tested on SQL Server.

Examples:
```C#
var petNames = context.People.SelectMany(x => x.Pets).ToList();
```
```
Unhandled exception. System.InvalidOperationException: SelectMany's collectionSelector was not NavigationExpansionExpression
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.ProcessSelectMany(NavigationExpansionExpression source, MethodInfo genericMethod, LambdaExpression collectionSelector, LambdaExpression resultSelector)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.Expand(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryOptimizer.Visit(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.Main() in C:\Stuff\ThreeOhP7P\ThreeOhP7P\Program.cs:line 335
   at Program.<Main>()
```

```C#
var petNames = context.People.SelectMany(x => x.Pets.Select(y => y.Name)).ToList();
```
```
Unhandled exception. System.ArgumentException: The type or method has 2 generic parameter(s), but 3 generic argument(s) were provided. A generic argument must be provided for each generic parameter.
   at System.RuntimeType.SanityCheckGenericArguments(RuntimeType[] genericArguments, RuntimeType[] genericParamters)
   at System.Reflection.RuntimeMethodInfo.MakeGenericMethod(Type[] methodInstantiation)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.ProcessSelectMany(NavigationExpansionExpression source, MethodInfo genericMethod, LambdaExpression collectionSelector, LambdaExpression resultSelector)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at System.Linq.Expressions.MethodCallExpression.Accept(ExpressionVisitor visitor)
   at System.Linq.Expressions.ExpressionVisitor.Visit(Expression node)
   at Microsoft.EntityFrameworkCore.Query.Internal.NavigationExpandingExpressionVisitor.Expand(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryOptimizer.Visit(Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass9_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQueryCore[TFunc](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Program.Main() in C:\Stuff\ThreeOhP7P\ThreeOhP7P\Program.cs:line 335
   at Program.<Main>()
```
Workaround: evaluate the SelectMany on the client:
```C#
var petNames = context.People
    .Include(e => e.Pets)
    .AsEnumerable()
    .SelectMany(x => x.Pets.Select(y => y.Name)).ToList();
```
Full repo:
```C#
public class Person
{
    public int Id { get; set; }
    public string Name { get; set; }
    public List<Pet> Pets { get; set; }
}

public class Pet
{
    public int Id { get; set; }
    public string Name { get; set; }
    public int PersonId { get; set; }
    public Person Person { get; set; }
}

public class BloggingContext : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .UseSqlServer(@"Server=(localdb)\mssqllocaldb;Database=Test;ConnectRetryCount=0");

    public DbSet<Person> People { get; set; }
}

public class Program
{
    public static async Task Main()
    {
        using (var context = new BloggingContext())
        {
            context.Database.EnsureDeleted();
            context.Database.EnsureCreated();

            context.Add(new Person
            {
                Name = "brock",
                Pets = new List<Pet>
                    {
                        new Pet
                        {
                            Name = "max"
                        }
                    }
            });
            context.Add(new Person
            {
                Name = "dom",
                Pets = new List<Pet>
                    {
                        new Pet
                        {
                            Name = "mittens"
                        }
                    }
            });

            context.SaveChanges();
        }

        using (var context = new BloggingContext())
        {
            var petNames = context.People.SelectMany(x => x.Pets.Select(y => y.Name)).ToList();
        }
    }
}
```

### PR DESCRIPTION
**Pull request by smitpatel**
SelectMany in linq look something like following (second clause is considered the collection selector)
```C#
from c in cs
from o in c.Orders
select....
```
Following are the translations of SelectMany

Unrelated collection selector
```C#
from c in cs
from o in os
```
Generates CROSS JOIN

Correlated collection selector with correlation being a predicate
```C#
from c in cs
from o in os.Where(o => o.CustomerID == c.CustomerID)
```
Such predicate can be lifted and used in generating a join. So this query generates JOIN.
If collection selector ends with DefaultIfEmpty then it is LEFT JOIN.

Correlated collection selector with correlation not a predicate
```C#
from c in cs
from o in os.Select(o => c.City)
```
Since we cannot generate a join predicate here, this translates to CROSS APPLY.
If collection selector ends with DefaultIfEmpty then it is OUTER APPLY.

- Add support for Cross Apply
- Add Support for Outer Apply
- Convert Cross Apply to Inner Join when possible
- Convert Outer Apply to Left Join when possible
- Add translation for DefaultIfEmpty
- Add translation for both overloads of SelectMany
- Handle DefaultIfEmpty & SelectMany without collectionSelector in Navigation Expansion
- Ensure columns are in projection when generating join predicate from a correlated subquery

Currently there are no tests for Cross/Outer Apply.
Our earlier Cross Apply got converted to join. N+1 evaluation tests are disabled right now, which would generate Cross Apply.
We never supported Outer Apply in past.

Resolves #15711
Resolves #12567
Resolves #12572
Resolves #12872
Resolves #16330
Resolves #15081
Resolves #16989

Re-enable tests for #12449

### CODE REVIEW

**Comment by ajcvickers** [comment]
@maumar Please make it a priority to review this when you get in. (You were nominated in triage! 🚎)

File: test/EFCore.SqlServer.FunctionalTests/Query/OwnedQuerySqlServerTest.cs [hunk 1]
```diff
@@ -225,16 +225,17 @@ public override void Query_when_subquery()
...
-LEFT JOIN [OwnedPerson] AS [o1] ON [o8].[Id] = [o1].[Id]
+LEFT JOIN [OwnedPerson] AS [o1] ON [o0].[Id] = [o1].[Id]
 LEFT JOIN [OwnedPerson] AS [o2] ON [t0].[Id] = [o2].[Id]
 LEFT JOIN [OwnedPerson] AS [o3] ON [o2].[Id] = [o3].[Id]
 LEFT JOIN [OwnedPerson] AS [o4] ON [t0].[Id] = [o4].[Id]
 LEFT JOIN [OwnedPerson] AS [o5] ON [o4].[Id] = [o5].[Id]
 LEFT JOIN [OwnedPerson] AS [o6] ON [t0].[Id] = [o6].[Id]
 LEFT JOIN [OwnedPerson] AS [o7] ON [o6].[Id] = [o7].[Id]
-ORDER BY [t0].[Id]");
+LEFT JOIN [Order] AS [o8] ON [t0].[Id] = [o8].[ClientId]
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/OwnedQuerySqlServerTest.cs]: why extra join here?

File: test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs [hunk 2]
```diff
@@ -6505,7 +6493,7 @@ public override async Task Cast_to_derived_type_after_OfType_works(bool isAsync)
             AssertSql(
                 @"SELECT [g].[Nickname], [g].[SquadId], [g].[AssignedCityName], [g].[CityOrBirthName], [g].[Discriminator], [g].[FullName], [g].[HasSoulPatch], [g].[LeaderNickname], [g].[LeaderSquadId], [g].[Rank]
 FROM [Gears] AS [g]
-WHERE [g].[Discriminator] = N'Officer'");
+WHERE [g].[Discriminator] IN (N'Gear', N'Officer') AND ([g].[Discriminator] = N'Officer')");
```
Comment (maumar) [inline test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs]: we should optimize this at some point (later)

File: test/EFCore.SqlServer.FunctionalTests/Query/OwnedQuerySqlServerTest.cs (hunk 1 above)
Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/OwnedQuerySqlServerTest.cs]: Orders was owned collection and it was ignored earlier.

File: test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs (hunk 2 above)
Comment (smitpatel) [reply test/EFCore.SqlServer.FunctionalTests/Query/GearsOfWarQuerySqlServerTest.cs]: You mean #15808 :trollface:

File: src/EFCore.SqlServer/Query/Internal/SearchConditionConvertingExpressionVisitor.cs [hunk 3]
```diff
@@ -317,6 +317,26 @@ protected override Expression VisitCrossJoin(CrossJoinExpression crossJoinExpres
             return crossJoinExpression.Update(table);
         }
 
+        protected override Expression VisitCrossApply(CrossApplyExpression crossApplyExpression)
```
Comment (maumar) [inline src/EFCore.SqlServer/Query/Internal/SearchConditionConvertingExpressionVisitor.cs]: logic repeated 3 times - dry

Comment (smitpatel) [reply src/EFCore.SqlServer/Query/Internal/SearchConditionConvertingExpressionVisitor.cs]: There is no good way to do it. Update function does not exist on abstract class.

**Comment by smitpatel** [comment]
Renamed CrossApplyExpression -> InnerJoinLateralExpression & OuterApplyExpression -> LeftJoinLateralExpression to match with formal SQL terminology. Also updated relational implementation and moved SqlServer specific sql syntax in provider code.

**Pull request merged**